lib/web_ui/test/golden_tests/engine/canvas_clip_path_test.dart

```diff
@@ -0,0 +1,105 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:html' as html;
+import 'dart:js_util' as js_util;
+
+import 'package:ui/ui.dart' hide TextStyle;
+import 'package:ui/src/engine.dart' as engine;
+import 'package:test/test.dart';
+
+import 'package:web_engine_tester/golden_tester.dart';
+
+void main() async {
+  const double screenWidth = 600.0;
+  const double screenHeight = 800.0;
+  const Rect screenRect = Rect.fromLTWH(0, 0, screenWidth, screenHeight);
+
+  // Commit a recording canvas to a bitmap, and compare with the expected
+  Future<void> _checkScreenshot(engine.RecordingCanvas rc, String fileName,
+      {Rect region = const Rect.fromLTWH(0, 0, 500, 500)}) async {
+    final engine.EngineCanvas engineCanvas = engine.BitmapCanvas(screenRect);
+
+    rc.apply(engineCanvas);
+
+    // Wrap in <flt-scene> so that our CSS selectors kick in.
+    final html.Element sceneElement = html.Element.tag('flt-scene');
+    try {
+      sceneElement.append(engineCanvas.rootElement);
+      html.document.body.append(sceneElement);
+      await matchGoldenFile('$fileName.png', region: region, maxDiffRate: 0.2);
+    } finally {
+      // The page is reused across tests, so remove the element after taking the
+      // Scuba screenshot.
+      sceneElement.remove();
+    }
+  }
+
+  setUp(() async {
+    debugEmulateFlutterTesterEnvironment = true;
+    await webOnlyInitializePlatform();
+    webOnlyFontCollection.debugRegisterTestFonts();
+    await webOnlyFontCollection.ensureFontsLoaded();
+  });
+
+  // Regression test for https://github.com/flutter/flutter/issues/48683
+  // Should clip image with oval.
+  test('Clips image with oval clip path', () async {
+    final engine.RecordingCanvas rc =
+        engine.RecordingCanvas(const Rect.fromLTRB(0, 0, 400, 300));
+    rc.save();
+    Image testImage = createTestImage();
+    double testWidth = testImage.width.toDouble();
+    double testHeight = testImage.height.toDouble();
+    final Path path = Path();
+    path.addOval(Rect.fromLTWH(100, 30, testWidth, testHeight));
+    rc.clipPath(path);
+    rc.drawImageRect(testImage, Rect.fromLTRB(0, 0, testWidth, testHeight),
+        Rect.fromLTWH(100, 30, testWidth, testHeight), Paint());
+    rc.restore();
+    await _checkScreenshot(rc, 'image_clipped_by_oval');
+  });
+
+  // Regression test for https://github.com/flutter/flutter/issues/48683
+  test('Clips triangle with oval clip path devicePixelRatio2', () async {
```

Comment:
What does "devicePixelRatio2" in the description mean? I wasn't able to parse it from the test itself.

Reply:
removed. Initially explored EngineWindow.debugDevicePixelRatio but snapshots still need some infra updates as discussed.